### PR TITLE
Require backward compatibility down to Java 8 

### PR DESCRIPTION
**Pull request by marcelluethi**

### CODE REVIEW

File: project/Build.scala [hunk 1]
```diff
@@ -12,7 +12,10 @@ object BuildSettings {
   val buildScalaVersion = "2.12.6"
 
   val buildSettings = Defaults.defaultSettings ++ Seq(
-    scalacOptions ++= Seq("-encoding", "UTF-8", "-Xlint", "-deprecation", "-unchecked", "-feature"),
+    scalacOptions ++= (CrossVersion.partialVersion(scalaVersion.value) match {
+      case Some((2,  11)) =>  Seq("-encoding", "UTF-8", "-Xlint", "-deprecation", "-unchecked", "-feature", "-target:jvm-1.6")
```
Comment (Andreas-Forster) [inline project/Build.scala:16]: Regarding our offline discussion, does it make sense to have support for Java 1.6?

**Review by Andreas-Forster — APPROVED** [review]
Looks fine to me. The only point I am unsure is if it makes sense to have backward compatibility to Java 1.6?

**Comment by clangguth** [comment]
@Andreas-Forster Let's put it the other way around: is there a good reason for **not** staying compatible with older JVM versions?

Unless old versions really get "too old", because they don't support something that is absolutely required, or because supporting them becomes a burden, I'm always in favor of being as downward-compatible as reasonably possible. 

Most people don't run the newest version of everything, so IMO: if there's no explicit need for something, then don't require it.

Just my two cents, though.

**Comment by Andreas-Forster** [comment]
@Christoph Langguth <christoph.langguth@unibas.ch> In general, I agree that
one could look from "the other side" at the problem. But personally, I
believe that faster migration to newer software should always be preferred
and is valuable. In addition, Java 8 is out for 5 years and at least the
freely available JVM 8 of Oracle does not get any updates anymore since
January. So already this version is quite old. For JVM 6 free updates were
stopped in 2013.

So my main questions if we really want to keep JVM 6 are:
- Has anybody run and tested our software using JVM 6?
- And what is the impact of using backward compatibility? Performance loss?
Other?

**Comment by clangguth** [comment]
At least while initiallly developing it, I have rather extensively tested the software on Java 6, 7 and 8 on all supported platforms.

There might be a minimal performance impact due to using "older" classes for some internal implementations (the only concrete example I know was the StringBuilder replacing StringBuffer for string concatenation in Java 5 vs. 4).

As far as I understand, Scala 2.12 requires a Java 8 runtime anyway (because of some new features like closures(?) or lambdas(?) or so). So everybody using Scala 2.12 will necessary need JRE >= 8. And since Scala 2.11 still supports JRE 6 (correct me if I'm wrong), so should we. IMO :-)

**Comment by marcelluethi** [comment]
@clangguth  .... and this is, what this PR does 😄

**Pull request merged**